src/ledger/insert_shred.zig

```diff
@@ -810,6 +873,12 @@ pub const ShredInserter = struct {
         data_index: *meta.ShredIndex,
     ) bool {
         const shred_index: u64 = @intCast(shred.fields.common.index);
+        // QUESTION: is the "shred_index < slot_meta.consecutive_received_from_0" check needed?
+        // Ie would there ever be a case where data_index.contains(shred_index) is true but
+        // shred_index < slot_meta.consecutive_received_from_0 is false? If not then is data_index.contains(shred_index)
+        // not enough check?
+        // Also if shred_index > slot_meta.consecutive_received_from_0 is true and also data_index.contains(shred_index) is true
+        // is it a valid case case for data_index.contains(shred_index) to be true but shred_index > slot_meta.consecutive_received_from_0.
         return shred_index < slot_meta.consecutive_received_from_0 or
             data_index.contains(shred_index);
```

Comment:
You made a great observation here. But there is actually a specific performance reason for this redundancy.

`data_index.contains(shred_index)` tells us the correct answer. This could be the only thing in the function, and it would return the correct answer.

`shred_index < slot_meta.consecutive_received_from_0`, on the other hand, is not perfect. If this is true, then `data_index.contains(shred_index)` will definitely be true. But If this is false, `data_index.contains(shred_index)` might be either true or false. This is because shreds do not always come in order.

So, why bother with the `<` comparison at all? The idea is that `<` is faster than `.contains()`. We run the fast check first. If true, that's a definitive answer. But if false, we can't know for sure, so we run the slower check to get a definitive answer.